### PR TITLE
Feature/9

### LINKED ISSUE #9: Authentication & authorization
Research how to implement this and decide which functionality is required, like: Lost password, refresh token and others.

### PR DESCRIPTION
**Pull request by KristianS93**
## Description

The purpose of this pull request is to implement token based RBAC via JWT's for instructors.
Initially the only part of the issue is a login end point, however most of the register endpoint is ready to be implemented. Also functionality for tokens for anonymous users are also created. 
The JWT will always contain:
- A userId in a claim called `UserData`, represented as an integer.
- A role in a claim called `Role`, represented as a string.

A anonymous user will always have a userId of `-1`, unsure if there are problems regarding having multiple users with this userid. The expiration time of this token, will be set to the amount of time the session is set to. 

**Considerations**
There are some consideration to be made, the approach used here uses a `users` table in the database instead of an instructor class, as this seems to just be wasting some time as this would have to be changed later anyway when creating users.

Testing:
- Should Repositories be tested? They should only contain SQL calls.
- Substitution / mocking, requires and interface or a empty constructor.

Register endpoint?

**Before merging**
Based on these discussions, some changes are expected to be made both in SQL schemas, tests and application code. 

### Resolved Issue
Fixes #9 #13

## Changes

- /login endpoint
- UserRepository
- UserService
- Multiple models
- Changes to the schema

## Checklist

- [X] I have added test cases for my additions
- [X] New and old tests passed
- [ ] Documentation is updated

### CODE REVIEW

File: Infrastructure/Authentication/TokenService.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    public string GenerateAnonymousUserJwt(int sessionLength)
+    {
+        var token = new JwtSecurityToken(
+            issuer: AuthConstants.Issuer,
+            audience: AuthConstants.Audience,
+            claims: new List<Claim>
+            {
+                new Claim(ClaimTypes.UserData, AuthConstants.AnonymousUserId.ToString()),
+                new Claim(ClaimTypes.Role, nameof(Roles.AnonymousUser))
```
Comment (JMyrtue) [inline Infrastructure/Authentication/TokenService.cs:41]: Add a sessionId claim

File: Infrastructure/Authentication/TokenService.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        );
+        return new JwtSecurityTokenHandler().WriteToken(token);
+    }
+
+    private SigningCredentials GetSigningCredentials()
+    {
+        var key = Environment.GetEnvironmentVariable(AuthConstants.JwtSecret);
+        if (key == null)
+        {
+            _logger.LogWarning("No jwt key environment variable found!");
```
Comment (JMyrtue) [inline Infrastructure/Authentication/TokenService.cs:54]: Potentially add specific exception, and handle it in GlobalExceptionHandler

File: Infrastructure/Authentication/UserRepository.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	public async Task<User?> GetUserDetailsByEmailAsync(string email)
+	{
+		using var con = await _connection.CreateConnectionAsync();
+		var query = """
+					SELECT users.id, users.email, users.password_hash AS passwordhash, role.name AS rolestr FROM users
+						JOIN user_role ON users.id = user_role.user_id
+						JOIN role ON user_role.role_id = role.id
+					    WHERE email = @email;
+					""";
+		var user = await con.QuerySingleAsync<User>(query, new { email });
```
Comment (JMyrtue) [inline Infrastructure/Authentication/UserRepository.cs:41]: Change from single query to fit the situation where a User has more than one role

File: Infrastructure/Persistence/Scripts/001-InitialMigrationSchema.sql [hunk 4]
```diff
@@ -58,4 +58,24 @@ CREATE TABLE
...
\ No newline at end of file
+    );
+
+-- Handle auth
+CREATE TABLE users (
+    id SERIAL PRIMARY KEY,
+    email VARCHAR(50) UNIQUE NOT NULL,
+    password_hash VARCHAR(255) NOT NULL
+);
+-- Consider created at / last login
```
Comment (JMyrtue) [inline Infrastructure/Persistence/Scripts/001-InitialMigrationSchema.sql:69]: Add this, and include it in the hashing of password

File: UnitTest/Infrastructure/Authentication/UserServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+
+namespace UnitTest.Infrastructure.Authentication;
+
+public class UserServiceTest
+{
+	// User == null og != null
+	[Fact]
+	public async Task LoginAsync_UserEqualNull_ShouldReturn_FailedResult()
```
Comment (JMyrtue) [inline UnitTest/Infrastructure/Authentication/UserServiceTest.cs:13]: After discussion: Add redundant interfaces, that only is used for testing.

**Review by JMyrtue — APPROVED** [review]
Review done physical - comments added along the review, for minor modifications.

File: Infrastructure/Authentication/TokenService.cs (hunk 2 above)
Comment (KristianS93) [reply Infrastructure/Authentication/TokenService.cs:64]: fixed

**Comment by KristianS93** [comment]
For now this issue closes, however a new issue is created for the dummy data, and the student table has to be revised.

**Pull request merged**